ciso/_ciso.pyx

```diff
@@ -40,14 +40,15 @@ def zslice(double[:, :, ::1] q,
 
     cdef double[:, ::1] q_iso = np.empty((M, L), dtype=np.float64)
 
-    for i in range(L):
-        for j in range(M):
-            q_iso[j, i] = mask_val
-            for k in range(N-1):
-                if (((p[k, j, i] < p0) and (p[k+1, j, i] > p0)) or
-                   ((p[k, j, i] > p0) and (p[k+1, j, i] < p0))):
-                    dp = p[k+1, j, i] - p[k, j, i]
-                    dp0 = p0 - p[k, j, i]
-                    dq = q[k+1, j, i] - q[k, j, i]
-                    q_iso[j, i] = q[k, j, i] + dq*dp0/dp
+    with nogil:
+        for i in range(L):
+            for j in range(M):
+                q_iso[j, i] = mask_val
+                for k in range(N-1):
+                    if (((p[k, j, i] < p0) and (p[k+1, j, i] > p0)) or
+                       ((p[k, j, i] > p0) and (p[k+1, j, i] < p0))):
+                        q_iso[j, i] = (q[k, j, i] +
+                                       (q[k+1, j, i] - q[k, j, i]) *  # dq
+                                       (p0 - p[k, j, i]) /  # dp0
+                                       (p[k+1, j, i] - p[k, j, i]))  # dp
     return np.array(q_iso)
```

Comment:
has anyone checked if this is the most efficient wayt o do this? usually asarray() makes a copy, which we don't need here.

not a huge deal, but I think we can avoid that copy, but doing:

around line 41:

```
out_arr = np.empty((M, L), dtype=np.float64)
cdef double[:, ::1] q_iso = out_arr
```

then, at the end, simply:
```
return out_arr
```

what that will do is create a numpy array, as usual. then assign a memoryview object that shares data with that array, so changes to the memoryview object will change the array. So we can simply return that array at the end.

the other option is to foro the memoryview altogether, and simply do

'''
 cdef np.ndarray[np.float64, ndim=2, order=c] q_iso = np.empty((M, L), dtype=np.float64)
'''

I'm a bit lost as to whether working with memoryviews rather than arrays is faster or slower in modern Cython...

-CHB



-CHB